### PR TITLE
Add a fade-in animation for the text selection controls.

### PR DESCRIPTION
**Pull request by mpcomplete**
BUG=https://github.com/flutter/flutter/issues/3938

### CODE REVIEW

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 1]
```diff
@@ -133,13 +139,15 @@ class TextSelectionOverlay implements TextSelectionDelegate {
...
     Overlay.of(context, debugRequiredFor: debugRequiredFor).insertAll(_handles);
+    _controller.forward(from: 0.0);
   }
 
   /// Shows the toolbar by inserting it into the [context]'s overlay.
   void showToolbar() {
     assert(_toolbar == null);
     _toolbar = new OverlayEntry(builder: _buildToolbar);
     Overlay.of(context, debugRequiredFor: debugRequiredFor).insert(_toolbar);
+    _controller.forward(from: 0.0);
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/text_selection.dart]: Presumably you need to dispose `_controller` somewhere.

Comment (mpcomplete) [inline packages/flutter/lib/src/widgets/text_selection.dart]: I added a stop() call when the handles are hidden. Though AFAICT it's not harmful if the controller keeps ticking after the overlay goes away.

Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_selection.dart]: Well, it'll burn the battery down...

**Comment by Hixie** [comment]
Test please. :-)

You should have at least one test that tries to show the toolbar then ends the test before the toolbar has completely faded in.

**Comment by mpcomplete** [comment]
@Hixie I added a test that the toolbar has an Opacity ancestor with opacity less than 1.0. Is that what you had in mind?

**Comment by Hixie** [comment]
I mean a test that is in its own testWidgets, that includes a toolbar, then tells it to show and steps through the animation, verifying that the opacity is fading in, but that then ends the test before pumping enough frames that the animation would have ended.

**Comment by mpcomplete** [comment]
OK, separated the test out. I tested the beginning, middle, and end of the animation. (Or is there some reason you want the test to end before it fully fades in?)

**Comment by Hixie** [comment]
I want the test to end while the animation is still ongoing so that it verifies that the animation controller is being correctly disposed.

**Comment by mpcomplete** [comment]
OK, the test ends before the animation is finished now. Do we detect leaks in tests?

File: packages/flutter/test/widget/input_test.dart [hunk 2]
```diff
@@ -382,4 +382,58 @@ void main() {
...
+    String testValue = 'abc def ghi';
+    enterText(testValue);
+    await tester.pumpWidget(builder());
+
+    // Tap the selection handle to bring up the "paste / select all" menu.
+    await tester.tapAt(textOffsetToPosition(tester, testValue.indexOf('e')));
+    await tester.pumpWidget(builder());
+    RenderEditableLine renderLine = findRenderEditableLine(tester);
+    List<TextSelectionPoint> endpoints = renderLine.getEndpointsForSelection(
+        inputValue.selection);
```
Comment (Hixie) [inline packages/flutter/test/widget/input_test.dart]: if there's a line break after the (, there should be one before the )

**Comment by Hixie** [comment]
Yeah, we check to make sure there's no running animations at the end of each test.

**Comment by Hixie** [comment]
I expected it to notice that _handleController and _toolbarController aren't disposed (you should call their dispose() method from your dispose() method). It worries me that the test didn't fail.

**Comment by mpcomplete** [comment]
TextSelectionOverlay.hide() is being called before the test ends, which stop()s the controllers.

**Comment by Hixie** [comment]
What is calling it?

**Comment by mpcomplete** [comment]
RawInputLineState.dispose

**Comment by Hixie** [comment]
Ah, interesting. That makes sense.

We should still call dispose on the controllers in the selection itself when it is disposed, just for correctness.

Anyway, LGTM. Thanks for the tests.

**Pull request merged**